pkg/nextgen/distdoc/parser.go

```diff
@@ -0,0 +1,184 @@
+	} else {
+		document.certificate = certificates[0]
+		return nil
+	}
+}
+
+// parseSigningTime parses, validates and sets the document signing time.
+func parseSigningTime(document *document, headers jws.Headers, _ *jws.Message) error {
+	if timeAsInterf, ok := headers.Get(signingTimeHeader); !ok {
+		return documentValidationError(missingHeaderErrFmt, signingTimeHeader)
+	} else if timeAsFloat64, ok := timeAsInterf.(float64); !ok {
+		return documentValidationError(invalidHeaderErrFmt, signingTimeHeader)
+	} else {
+		document.signingTime = time.Unix(int64(timeAsFloat64), 0)
+		return nil
+	}
+}
+
+// parseVersion parses, validates and sets the document format version.
+func parseVersion(document *document, headers jws.Headers, _ *jws.Message) error {
+	var version Version
+	if versionAsInterf, ok := headers.Get(versionHeader); !ok {
+		return documentValidationError(missingHeaderErrFmt, versionHeader)
+	} else if versionAsFloat64, ok := versionAsInterf.(float64); !ok {
+		return documentValidationError(invalidHeaderErrFmt, versionHeader)
+	} else if version = Version(versionAsFloat64); version != currentVersion {
+		return documentValidationError("unsupported version: %d", version)
+	} else {
+		document.version = version
+		return nil
+	}
+}
+
+// parsePrevious parses, validates and sets the document prevs fields.
+func parsePrevious(document *document, headers jws.Headers, _ *jws.Message) error {
+	if prevsAsInterf, ok := headers.Get(previousHeader); !ok {
+		return documentValidationError(missingHeaderErrFmt, previousHeader)
+	} else if prevsAsSlice, ok := prevsAsInterf.([]interface{}); !ok {
+		return documentValidationError(invalidHeaderErrFmt, previousHeader)
+	} else {
+		for _, prevAsInterf := range prevsAsSlice {
+			if prevAsString, ok := prevAsInterf.(string); !ok {
+				return documentValidationError(invalidHeaderErrFmt, previousHeader)
+			} else if prev, err := model.ParseHash(prevAsString); err != nil {
+				return documentValidationError(invalidHeaderErrFmt, previousHeader)
+			} else {
+				document.prevs = append(document.prevs, prev)
+			}
+		}
+		return nil
+	}
+}
+
+// parseTimelineID parses, validates and sets the document timeline ID field.
+func parseTimelineID(document *document, headers jws.Headers, _ *jws.Message) error {
+	if tidAsInterf, _ := headers.Get(timelineIDHeader); tidAsInterf != nil {
+		if tidAsString, ok := tidAsInterf.(string); !ok {
+			return documentValidationError(invalidHeaderErrFmt, timelineIDHeader)
+		} else if timelineID, err := model.ParseHash(tidAsString); err != nil {
+			return documentValidationError(invalidHeaderErrFmt+": %w", timelineIDHeader, err)
+		} else {
+			document.timelineID = timelineID
+		}
+	}
+	return nil
+}
+
+// parseTimelineVersion parses, validates and sets the document timeline version field. Timeline ID must be present when
+// timeline version is present.
+func parseTimelineVersion(document *document, headers jws.Headers, _ *jws.Message) error {
```

Comment:
If the specs are unclear they should be revised (I'll review them).

The general idea is that `tiv=1` is used for the first update (so kinda based 0). It would feel semantically correct  to let the creating document `tiv=1` and the first update `tiv=2`, but since it's an optional field you could leave out for the first document and then the first update would have `tiv=2` without a `tiv=1` in the timeline, which is even stranger imo. Any ideas?

Reply:
if no tid is present, current document hash is presumed
if no tiv is present, 0 is presumed
?

Reply:
> I might have missed it, but:
> 
> * where does certificate validation happen? §3.6 of the rfc
Not yet implemented yet, because I wanted to keep this RFC a bit compact.
> * where does timeline validation happen? §3.4 of the rfc
Will implement this according to https://github.com/nuts-foundation/nuts-specification/pull/8 so this PR stays a bit compact.
>   I believe we talked about having those validations happen at a higher level, in that case: add some issues and maybe some references in the parser about validations happen at a different architectural level?
Detailed validation in a change to RFC004: https://github.com/nuts-foundation/nuts-specification/pull/8

Made issues to keep track;
- https://github.com/nuts-foundation/nuts-network/issues/45
- https://github.com/nuts-foundation/nuts-network/issues/46

Reply:
I like your suggestion, i'll update it.